framework/src/materials/Material.C

```diff
@@ -182,12 +187,45 @@ Material::getOutputs()
 }
 
 void
+Material::subdomainSetup()
+{
+  if (_constant_option == 2)
+  {
+    _current_subdomain_id = _current_elem->subdomain_id();
+
+    unsigned int nqp = _fe_problem.getMaxQps();
+
+    MaterialProperties & props = _material_data->props();
+    for (const auto & prop_id : _supplied_prop_ids)
+      props[prop_id]->resize(nqp);
+
+    _qp = 0;
+    computeSubdomainProperties();
+
+    for (const auto & prop_id : _supplied_prop_ids)
+    {
+      for (decltype(nqp) qp = 1; qp < nqp; ++qp)
+        props[prop_id]->qpCopy(qp, props[prop_id], 0);
+    }
+  }
+}
+
+void
+Material::computeSubdomainProperties()
+{
+  mooseError("computeSubdomainQpProperties in Material '", name(), "' needs to be implemented");
+}
+
+void
 Material::computeProperties()
 {
+  if (_constant_option == 2)
```

Comment:
I mean "magic" as in an undocumented constant. Yes I can see that it's in the MOOSE Enum. Lately, we've gotten into the habit of creating a class scoped enum that we dump the MooseEnum into so that it's clean. Take a look at how this works:
https://github.com/idaholab/moose/blob/devel/framework/src/postprocessors/NumDOFs.C#L35